A29-xds-tls-security.md

```diff
@@ -0,0 +1,509 @@
+using the client's own configuration instead of being sent by the control plane. This enables flexibility
+in heterogeneous deployments where different clients can use different implementations for the same
+provider instance name. For example, an on-prem client may get its certificate using a local certificate
+provider whereas a client running in cloud may use a cloud-vendor provided implementation to obtain
+the certificate.
+
+gRPC provides a `CertificateProvider` plugin API that can support multiple implementations. The plugin API
+is not currently public, so applications cannot currently add their own provider implementations although
+we might make this API public in the future. However gRPC currently includes a `file_watcher` provider
+implementation - descibed [below][FILE-WATCHER-LINK] - which reads certificates from the local file system.
+
+[FILE-WATCHER-LINK]: #file_watcher-certificate-provider
+
+Certificate provider plugin instances are configured via the xDS bootstrap file.  There is a new
+top-level field called `"certificate_providers"` whose value is a map (a JSON object).  The key
+in this map is the certificate_provider_instance name and the value is a JSON object
+having exactly 2 fields:
+
+* `"plugin_name"` which is the name of the plugin implementation (a string value), and
+* `"config"` which is the configuration for the plugin. The value of `"config"` is a JSON object
+whose schema is defined by that plugin.
+
+The structure in the bootstrap file is as follows:
+
+```
+"certificate_providers": {
+  "instance_name": {
+    "plugin_name": "implementation_name",
+    "config": {
+      // ...config for implementation_name plugin...
+    }
+  }
+}
+```
+
+This defines a plugin instance for the instance name `instance_name` which consists of a
+plugin implementation called `implementation_name` with the corresponding configuration
+defined as the value for the key `config`. When the xDS server tells the client to obtain
+a certificate from `instance_name`, the client will use this plugin instance.
+
+#### `file_watcher` Certificate Provider
+
+As mentioned before, we currently only have one plugin implementation called `file_watcher`.
+The configuration for this plugin has the following fields:
+- `certificate_file`: The path to the file containing the identity certificate. The file should contain
+a PEM-formatted X.509 conforming certificate.
+- `private_key_file`: The path to the file containing the private key. The file should contain a
+PEM-formatted PKCS encoded private key.
+- `ca_certificate_file`: The path to the file containing the root certificate. The file should contain
+a PEM-formatted X.509 conforming certificate chain.
+- `refresh_interval`: Specifies how frequently the plugin should read the files. The value must be
+in the [JSON format described for a `Duration`][DURATION-JSON] protobuf message.
+
+For example, the bootstrap file might contain the following:
+
+```
```

Comment:
```suggestion
```jsonc
```

```jsonc
  "certificate_providers": {
    "google_cloud_private_spiffe": { // certificate_provider_instance name
      "plugin_name": "file_watcher", // name of the plugin instance
      "config": {                    // config to be supplied to the plugin instance
        "certificate_file": "/var/run/gke-spiffe/certs/certificates.pem",
        "private_key_file": "/var/run/gke-spiffe/certs/private_key.pem",
        "ca_certificate_file": "/var/run/gke-spiffe/certs/ca_certificates.pem",
        "refresh_interval": "60s"
      }
    }
  }
```